src/main/java/com/redhat/devtools/lsp4ij/features/documentSymbol/LSPDocumentSymbolUtils.java

```diff
@@ -0,0 +1,127 @@
+ * Copyright (c) 2025 Red Hat Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.documentSymbol;
+
+import com.intellij.ide.structureView.StructureViewTreeElement;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPDocumentSymbolViewElement;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPFileStructureViewElement;
+import com.redhat.devtools.lsp4ij.features.semanticTokens.viewProvider.LSPSemanticTokensFileViewProvider;
+import org.eclipse.lsp4j.DocumentSymbol;
+import org.eclipse.lsp4j.Range;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Utility class for working with document symbols.
+ */
+final class LSPDocumentSymbolUtils {
+
+    private LSPDocumentSymbolUtils() {
+        // Pure utility class
+    }
+
+    /**
+     * Returns the structure view model for the provided element.
+     *
+     * @param element the element
+     * @return the structure view model for the element, or null if none was found
+     */
+    @Nullable
+    static LSPDocumentSymbolStructureViewModel getStructureViewModel(@NotNull PsiElement element) {
+        Editor editor = LSPIJUtils.editorForElement(element);
+        if (editor != null) {
+            PsiFile file = element.getContainingFile();
+            return new LSPDocumentSymbolStructureViewModel(file, editor);
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the closest containing document symbol data for the element.
+     *
+     * @param element the element
+     * @return the closest containing document symbol data, or null if none was found
+     */
+    @Nullable
+    static DocumentSymbolData getDocumentSymbolData(@NotNull PsiElement element) {
+        LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider = LSPSemanticTokensFileViewProvider.getInstance(element);
+        int effectiveOffset = semanticTokensFileViewProvider != null ? semanticTokensFileViewProvider.getEffectiveOffset(element) : -1;
```

Comment:
`effectiveOffset` here is the offset that was requested if not on an actual element or the offset of the requested actual element. Again, this is part of how we can handle interpolation between real elements and show the correct breadcrumbs when the caret is on an offset that doesn't correspond to a (non-file) PSI element and sticky lines as the user scrolls the editor.